include/cuco/utility/helper_functions.hpp

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/detail/error.hpp>
+
+#include <cstddef>
+
+namespace cuco {
+
+/// The default occupancy percentage: 0.5 implies a 50% occupancy
+constexpr double DEFAULT_HASH_TABLE_OCCUPANCY = 0.5;
+
+/**
+ * @brief Computes requisite size of hash table
+ *
+ * Computes the number of entries required in a hash table to satisfy
+ * inserting a specified number of keys to achieve the specified hash table
+ * occupancy.
+ *
+ * @note The output of this helper function for a static extent is not a build-time constant.
+ *
+ * @throw If the desired occupancy is no bigger than zero
+ * @throw If the desired occupancy is larger than one
+ *
+ * @param size The number of elements that will be inserted
+ * @param desired_occupancy The desired occupancy percentage, e.g., 0.5 implies a
+ * 50% occupancy
+ *
+ * @return The size of the hash table that will satisfy the desired occupancy for the specified
+ * input size
+ */
+constexpr std::size_t compute_hash_table_size(
+  std::size_t size, double desired_occupancy = DEFAULT_HASH_TABLE_OCCUPANCY)
+{
+  CUCO_EXPECTS(desired_occupancy > 0., "Desired occupancy must be larger than zero");
+  CUCO_EXPECTS(desired_occupancy <= 1., "Desired occupancy cannot be larger than one");
+
+  // Calculate size of hash map based on the desired occupancy
+  return static_cast<std::size_t>(static_cast<double>(size) / desired_occupancy);
```

Comment:
I'm all for DRY and all, but do we really need this?

Reply:
I'm not sure if we "really" need it as a public API or it should be an internal cuco API, but in cudf we call it all the time whenever we create a set/map:
```
 auto set = cuco::experimental::static_map/set{
      cuco::experimental::extent{compute_hash_table_size(input.num_rows())},....
```

Reply:
It's a handy tool that allows users to have easy occupancy control. libcudf uses it every time a cuco map/set is involved and I assume other users could benefit from this helper function as well.

If we think from a library perspective, none of the standard libraries like STL, boost or libcuc++ provide such a helper utility and I guess that's why @jrhemstad is reluctant to add it in cuco?

Reply:
IMO the users even don't need to be exposed to such utility, low level API. Ideally, they just call this:
```
auto set = cuco::static_set(input.num_rows());
```

I.e., they don't need to specify anything else. No occupancy control, no allocator etc.

Reply:
I'm reluctant just because it's kinda awkward. Like @ttnghia said, it's the kind of thing that should just be part of a constructor. Maybe add another constructor overload that takes a size and load factor. Or maybe add a defaulted load factor to the exist ctor.

Reply:
OK, I will leave the utility in libcudf then.

Having a default load factor or a constructor overload is not a thing in STL or boost hash containers.

Reply:
That's largely because cuco is special for the fact that the container can't grow. You'd want to look for inspiration from similar data structures like folly::AtomicHashArray: https://github.com/facebook/folly/blob/64c8e8d11b4832fcea45a115cdfb04ad5c92aa8d/folly/AtomicHashArray.h#L153-L176

Reply:
> Having a default load factor or a constructor overload is not a thing in STL or boost hash containers.

Not only that, but it would also be a footgunny API change: Without user intervention, after pulling this change, all cuco maps/sets will suddenly allocate twice the size (if we assume a default load factor of 50%).

We can discuss what the `size` argument in the ctor should stand for.
Currently, we assume it is a `capacity`, i.e., the total number of slots the container can hold. Well, not exactly, if we consider that the capacity needs to be prime, etc…
If, instead, the `size` argument stands for "the number of elements the container can at least store", and then add an additional tuning knob for the target occupancy (either via a ctor argument or config struct), I'm not opposed.

One additional thing: 
I think a default occupancy of 50% is debatable. Sure, 50% gives you great performance out-of-the-box, but one of the major advantages of using a hash table lies in its comparable small memory footprint. With 50% occupancy, we need 2N memory to store N keys. That's the same footprint a naive sort-and-compress workflow would use while being potentially faster.

Reply:
>  Without user intervention, after pulling this change, all cuco maps/sets will suddenly allocate twice the size

Right, so instead of adding a default load factor parameter, I'm working on adding a ctor overload taking the desired load factor as input. As a result, the `capacity` input is updated as `n` which stands for the number of elements to insert into the container. One small caveat is that this ctor cannot keep static extent "static" anymore since load factor is a runtime parameter.

Reply:
> One small caveat is that this ctor cannot keep static extent "static" anymore since load factor is a runtime parameter.

Right. We should add a check that the input extent is indeed dynamic.

Reply:
What's wrong with calling `static_set(num_elements/target_occupancy, …)`? The benefit of this ctor/helper isn't 100% clear to me 🙃

Reply:
> We should add a check that the input extent is indeed dynamic.

And that's not easy with the current code. I will probably backlog this PR for now.